modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -107,24 +120,30 @@ public void onTopologyLeap(LogicalTopologySnapshot newTopology) {
         }
     };
 
+    /** The logical topology on the last watch event. */
+    private Set<String> logicalTopology = Collections.emptySet();
```

Comment:
This variable should be thread safe, it's possible that you will init in one thread(node startup thread) and will try to read in watch thread, thus visibility is broken.

Reply:
fixed